src/freenet/node/diagnostics/diagnostics/threads/ThreadsDiagnostics.java

```diff
@@ -0,0 +1,141 @@
+    private static final int MONITOR_INTERVAL = 10000;
+    private static final String MONITOR_THREAD_NAME = "NodeDiagnostics: thread monitor";
+
+    /** Sleep interval to calculate % CPU used by each thread */
+    private static final int CPU_SLEEP_INTERVAL = 1000;
+    List<NodeThreadInfo> nodeThreadInfos    = Collections.synchronizedList(new ArrayList<>());
+
+    OperatingSystemMXBean operatingSystemMXBean = ManagementFactory.getOperatingSystemMXBean();
+    ThreadMXBean threadMxBean               = ManagementFactory.getThreadMXBean();
+    RuntimeMXBean runtimeMxBean             = ManagementFactory.getRuntimeMXBean();
+
+    /**
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     * @param name Thread name
+     * @param monitorInterval Sleep intervals to retrieve CPU usage
+     */
+    public ThreadsDiagnostics(NodeStats nodeStats, Ticker ticker, String name, int monitorInterval) {
+        this.nodeStats = nodeStats;
+        this.ticker = ticker;
+        this.name = name;
+        this.monitorInterval = monitorInterval;
+    }
+
+    /*
+     * @param nodeStats Used to retrieve data points
+     * @param ticker Used to queue timed jobs
+     */
+    public ThreadsDiagnostics(NodeStats nodeStats, Ticker ticker) {
+        this(nodeStats, ticker, MONITOR_THREAD_NAME, MONITOR_INTERVAL);
+    }
+
+    private void scheduleNext(int interval) {
+        ticker.queueTimedJob(
+            this,
+            name,
+            interval,
+            false,
+            true
+        );
+    }
+
+    public void start() {
+        scheduleNext(0);
+    }
+
+    private void scheduleNext() {
+        scheduleNext(monitorInterval);
+    }
+
+    @Override
+    public void run() {
+        Map<Long, Long> threadCPU = new HashMap<>();
+
+        for (ThreadInfo info : threadMxBean.dumpAllThreads(false, false)) {
+            threadCPU.put(
+                info.getThreadId(),
+                threadMxBean.getThreadCpuTime(
+                    info.getThreadId()
+                )
+            );
+        }
+
+        long initialUptime = runtimeMxBean.getUptime();
+
+        try {
+            Thread.sleep(CPU_SLEEP_INTERVAL);
+        } catch (InterruptedException ignored) {
+        }
+
+        for (ThreadInfo info : threadMxBean.dumpAllThreads(false, false)) {
+            Long prev = threadCPU.get(info.getThreadId());
+            if (prev == null) {
+                continue;
+            }
+
+            threadCPU.put(
+                info.getThreadId(),
+                threadMxBean.getThreadCpuTime(info.getThreadId()) - prev
+            );
+        }
+
+        long elapsedUptime = runtimeMxBean.getUptime() - initialUptime;
+        double totalElapsedUptime = (elapsedUptime * operatingSystemMXBean.getAvailableProcessors()) * 10000F;
+
+        List<NodeThreadInfo> threads = new ArrayList<>();
+        for (Thread t : nodeStats.getThreads()) {
```

Comment:
(Silly nitpicking, free free to ignore: single-char variable names should be constrained to when you, for whatsoever reason, need to run something 5 times or so; if you iterate over something that has a clearer purpose, I would appreciate a more appropriate name.)